scripts/lldb/firestore.py

```diff
@@ -0,0 +1,250 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import ast
+import json
+
+"""
+LLDB type summary providers for common Firestore types.
+
+This is primarily useful for debugging Firestore internals. It will add useful
+summaries and consolidate common types in a way that makes them easier to
+observe in the debugger.
+
+Use this by adding the following to your ~/.lldbinit file:
+
+    command script import ~/path/to/firebase-ios-sdk/scripts/lldb/firestore.py
+
+Most of this implementation is based on "Variable Formatting" in the LLDB online
+manual: https://lldb.llvm.org/use/variable.html. There are two major features
+we're making use of:
+
+  * Summary Providers: these are classes or functions that take an object and
+    produce a (typically one line) summary of the type
+
+  * Synthetic Children Providers: these are classes that provide an alternative
+    view of the data. The children that are synthesized here show up in the
+    graphical debugger.
+"""
+
+
+class ForwardingSynthProvider(object):
+  """A synthetic child provider that forwards all methods to another provider.
+
+  Override the `delegate` method to customize the target to which this forwards.
```

Comment:
Does using this class on its own (without overriding `delegate`) ever make sense? If no, perhaps `delegate` should throw rather than have a default behavior?

Reply:
I've almost convinced myself that a forwarding synth provider is a bad idea (it masks the actual path to the variables). I'd prefer not to invest much more in it until I'm more certain this will stick around.

Using this in its default configuration does no harm; it's just a no-op.